fetch/fetch.go

```diff
@@ -0,0 +1,506 @@
+func (f *Fetch) requestHashBatchFromPeers() {
+	var requestList []requestMessage
+	f.activeReqM.RLock()
+	for hash, req := range f.activeRequests {
+		requestList = append(requestList, requestMessage{Hash: hash, Hint: req[0].hint})
+	}
+	f.activeReqM.RUnlock()
+
+	// send in batches
+	for i := 0; i < len(requestList); i += f.cfg.BatchSize {
+		j := i + f.cfg.BatchSize
+		if j > len(requestList) {
+			j = len(requestList)
+		}
+		f.sendBatch(requestList[i:j])
+	}
+}
+
+// sendBatch dispatches batched request messages
+func (f *Fetch) sendBatch(requests []requestMessage) {
+	// build list of batch messages
+	var batch requestBatch
+	batch.Requests = requests
+	batch.SetID()
+
+	f.activeBatchM.Lock()
+	f.activeBatches[batch.ID] = batch
+	f.activeBatchM.Unlock()
+	// timeout function will be called if no response was received for the hashes sent
+	timeoutFunc := func(err error) {
+		f.handleHashError(batch.ID, err)
+	}
+
+	bytes, err := types.InterfaceToBytes(batch)
+	if err != nil {
+		f.handleHashError(batch.ID, err)
+	}
+	// try sending batch to some random peer
+	retries := 0
+	for {
+		if f.stopped() {
+			return
+		}
+		// get random peer
+		p := f.net.GetRandomPeer()
+		err := f.net.SendRequest(fetch, bytes, p, f.receiveResponse, timeoutFunc)
+		// if call succeeded, continue to other requests
+		if err != nil {
+			retries++
+			if retries > f.cfg.MaxRetiresForPeer {
+				f.handleHashError(batch.ID, ErrCouldNotSend(fmt.Errorf("could not send message")))
+				break
+			}
+			//todo: mark number of fails per peer to make it low priority
+			f.log.Warning("could not send message to peer %v, retrying, retries %v", p, retries)
+		} else {
+			break
+		}
+
+	}
+}
+
+// handleHashError is called when an error occurred processing batches of the following hashes
+func (f *Fetch) handleHashError(batchHash types.Hash32, err error) {
+	f.log.Error("cannot fetch message %v", err)
+	f.activeBatchM.RLock()
+	batch, ok := f.activeBatches[batchHash]
+	if !ok {
+		f.activeBatchM.RUnlock()
+		f.log.Error("batch invalidated twice %v", batchHash)
+	}
+	f.activeBatchM.RUnlock()
+	f.activeReqM.Lock()
+	for _, h := range batch.Requests {
+		for _, callback := range f.activeRequests[h.Hash] {
+			callback.returnChan <- HashDataPromiseResult{
+				Err:  err,
+				Hash: h.Hash,
+				Data: nil,
+			}
+		}
+		delete(f.activeRequests, h.Hash)
+	}
+	f.activeReqM.Unlock()
+
+	f.activeBatchM.Lock()
+	delete(f.activeBatches, batchHash)
+	f.activeBatchM.Unlock()
+}
+
+// HashDataPromiseResult is the result strict when requesting data corresponding to the Hash
+type HashDataPromiseResult struct {
+	Err  error
+	Hash types.Hash32
+	Data []byte
+}
+
+// GetAllHashes gets a list of hashes and validates them with the dataCallback function
+// it will return a map of hashes and their respective promise channels
+func (f *Fetch) GetAllHashes(hashes []types.Hash32, hint Hint, validateHash bool) map[types.Hash32]chan HashDataPromiseResult {
```

Comment:
Why are you calling it `GetAllHashes` and not `GetHashes`? I could understand the first in the prior version, where error was to be returned when any of them failed, but that's not the case anymore.